man/ffpfilt.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpfilt 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpfilt - This program filters FFP profiles based on numeric/statistical properties.
.SH SYNOPSIS
.BI "ffpfilt [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
This utility can be used to filter high or low frequency features
from the FFP matrix. The upper and/or lower limits can be set 
using a raw frequency cutoff or a probability cutoff which
assumes either a normal or an extreme value distribution. 
Requires key-valued FFP input.  Note you should use this as a replacement for 
.B ffpcol 
in any pipelines (rather than in addition to) as 
.B ffpfilt 
produces columnar output.  Columnar output can be disabled with the
.B --keys
option. FFPs will be read from standard input if no options are supplied and 
.B ffpfilt 
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list.  The default input is assumed to be nucleotide sequence.


.SH OPTIONS
.TP
.B -s, --stats
.PP
Print statistics describing the FFP frequency distribution to
standard error.
.TP 
.B -z, --zeros
.PP
Include zeros is calculation of statistical parameters. When few
features are observed, but a long feature length is used this 
has the effect of skewing the feature distribution towards zero.
.TP
.BI "-u " "FLOAT, " "--upper=" "FLOAT"
.PP
.RI "The upper threshold limit for raw frequencies where " "FLOAT" " can be "
.RI "any frequency value or for probability distributions, in which case" " FLOAT "
can range from 0 to 1.
.TP
.BI "-l " "FLOAT, " "--lower=" "FLOAT"
.PP
The lower threshold limit.
.TP
.B -f, --freq
.PP
Use raw frequencies for filtering
.TP
.B -n, --norm
.PP
Use a normal distribution for filtering
.TP
.B -e, --evd
.PP
Use an extreme value (Gumbel) distribution for filtering
.TP 
.B -d, --disable
.PP
Disable classing of amino acid and nucleotide sequences. 
.TP
.B -a, --amino
.PP
Input FFPs are animo acid sequences.
.TP
.B -t, --text
.PP
Input FFPs are text.
.TP
.B -k, --keys
.PP
Force key,value output.  Default it columnar output.
.TP
.B -h, --help
.PP
Display short help message.
.TP
.B -v, --version
.PP
Display version information.
.PP
.SH EXAMPLES
The main usage of this progam is for filtering out high frequency
features, for example large numbers of prophage copies in a genome
might be need to be filtered out -- since they don't provide much
useful phylogenetic signal.  Here we filter out all features which
occur more than 5 times using the 
.B \-u
option.
.PP
.CODE ffpry -l 5 *.fna > vectors
.CODE ffpfilt -f -u 5 vectors > vectors.filt
.PP
Features which occur rarely and features which occur frequently can
be removed simultaneously.  This example excludes all features which
occur more than 10 times and less than 2 times.
.PP
.CODE ffpfilt -f -u 10 -u 2
.PP
Finally we can assume that the feature frequencies follow a specific
distribution.  Usually the extreme value distribution [EVD] is the appropriate choice,
however you should confirm what the distribution of your feature frequencies are
yourself.  We can exclude features that lie outside a certain range of the cummulative
probability distribution.  Here we exclude all features which lie below
10% and above 95% of the EVD cummulative distribution function.
.PP
.CODE ffpfilt -e -l 0.1 -u 0.95
.PP
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpaa(1),
.BR ffpry(1),
.BR ffprwn(1),
.BR ffpjsd(1),
.BR ffpcol(1)



NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpfilt.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <math.h>
#include <getopt.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hash.h"
#include "utils.h"
#include "vstring.h"
#include "string.h"
#include "cdfmacros.h"
#include "sighandle.h"
#include "../config.h"

char PROG_NAME[FILENAME_MAX];
#define MAX_WORD_LENGTH 40 /**< Maximum feature length */
#define DEFAULT_PRECISION 2

void hashCol(FILE * fp);
int isKeyBased(FILE * fp);
int getKeyLength(FILE * fp);


char usage_str[] = "Usage: %s [OPTIONS]... [FILE]...\n\
This program filters FFP profiles based on numeric properties\n\n\
\t-s, --stats\tPrint stats to stderr\n\
\t-z, --zeros\tInclude zeros in statistics calculations\n\
\t-u FLOAT, --upper=FLOAT\tupper threshold limit\n\
\t-l FLOAT, --lower=FLOAT\tlower threshold limit\n\
\t-f, --freq\tExclude raw frequencies outside of -u and -l limits\n\
\t-n, --norm\tUse Normal CDF to exclude features out -u and -l probabilities\n\
\t-e, --evd\tUse Extreme value CDF to exclude features out -u and -l probabilities\n\
\t-d, --disable\tDisable classing of AAs and NAs\n\
\t-a, --amino\tInput is an amino acid sequence\n\
\t-t, --text\tInput is a text file\n\
\t-k, --key\tForce key output.\n\
\t-v, --version\tPrint version information\n\
\t-h, --help\tThis text.\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


char *weightVector;

int Length = MAX_WORD_LENGTH;
			    /**< Feature length */
/**< It might be clearer if a distribution mode enum were used */

bool sflag = false;
bool zflag = false;
bool fflag = false;
bool nflag = false;
bool dflag = false;
bool aflag = false;
bool cflag = false;
bool tflag = false;
bool eflag = false;
bool lflag = false;
bool uflag = false;
bool kflag = false;

int pvalue = DEFAULT_PRECISION;
float upper = INT_MAX;
float lower = INT_MIN;
int main(int argc, char **argv)
{
    FILE *fp = NULL;
    int opt;
    int mode = nucleotide;
    int option_index = 0;

    static struct option long_options[] = {
	{"help", no_argument, 0, 'h'},
	{"print-stats", no_argument, 0, 's'},
	{"version", no_argument, 0, 'v'},
	{"precision", required_argument, 0, 'p'},
	{"upper", required_argument, 0, 'u'},
	{"lower", required_argument, 0, 'l'},
	{"freq", no_argument, 0, 'f'},
	{"normal", no_argument, 0, 'n'},
	{"evd", no_argument, 0, 'e'},
	{"zeros", no_argument, 0, 'z'},
	{"disable", no_argument, 0, 'd'},
	{"keys", no_argument, 0, 'k'},
	{"amino", no_argument, 0, 'a'},
	{"text", no_argument, 0, 't'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();

    strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "svfnu:dhl:etak",
			      long_options, &option_index)) != -1)
	switch (opt) {
	case 's':		// print stats to stderr
	    sflag = !sflag;
	    break;
	case 'z':		// include zeros in stat calc
	    zflag = !zflag;
	    break;
	case 'l':
	    lflag = !lflag;
	    lower = atof(optarg);
	    break;
	case 'u':
	    uflag = !uflag;
	    upper = atof(optarg);
	    break;
	case 'e':		// evd filtering two args lower upper
	    eflag = !eflag;
	    break;
	case 'n':		// norm filtering two args lower upper
	    nflag = !nflag;
	    break;
	case 'f':		// freq filtering two args lower upper 
	    fflag = !fflag;
	    break;
	case 't':
	    tflag = !tflag;
	case 'a':
	    aflag = !aflag;
	    break;
	case 'd':
	    dflag = !dflag;
	    break;
	case 'k':
	    kflag = !kflag;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    break;
	}

    if ((argc - optind) > 1)
	fatal_msg("Specify only one File argument\n");

    if (eflag + nflag + fflag > 1)
	fatal_msg("Specify -f, -n or -e, but not more than 1\n");


    if (aflag)
	mode = amino;

    if (tflag)
	mode = text;


    initHash(0, mode, !dflag);
    struct stat input;


    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
		fatal_msg("%s: %s.\n", *argv,strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	if (!isRegularFile(fp)) 
	    fp = convertPipeToFile(fp);

	if (!isKeyBased(fp))
	    fatal_msg("%s: Not a key valued FFP.", *argv);

	Length = getKeyLength(fp);
	hashCol(fp);

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);


    return EXIT_SUCCESS;
}


/**
 * Convert a (key,value) FFP to a columnar FFP
 *
 * FFP in fp must be columnar FFP.
 *
 * @param fp A file pointer to a (key,value) FFP
 * @return none
 */



void hashCol(FILE * fp)
{
    unsigned val;
    char s[Length];
    char **keys;
    unsigned i;
    char c[2];
    int d;
    double x1 = 0;
    double x2 = 0;
    double sigma;
    long unsigned int ctr = 0;
    float alpha, beta;
    float zscore;
    int num;
    int row = 0;
    bool deleteKey;

    // Find all keys in the file.

    while (!feof(fp)) {
	d = fscanf(fp, "%s %u%*[\t]%[\n]", s, &val, c);
	s[Length] = '\0';
	//calc stats here.
	x1 += val;
	x2 += val * val;
	hashMax(s, val);	// Add to hash if greater than existing val
	ctr++;
	if (d == 3) {
	    row++;
	}
    }

    if (zflag)
	ctr = pow(2, Length) * row;	// if RY coded

    x1 /= ctr;
    x2 /= ctr;

    sigma = sqrt(x2 - x1 * x1);

    beta = sigma / PI_SQRT_6;
    alpha = x1 - beta * EULER_MASCHERONI;


    if (sflag)
	fprintf(stderr, "%.*e\t%.*e%.*e%.*e\n", pvalue, x1, pvalue, sigma,
		pvalue, alpha, pvalue, beta);


    hashKeys(&keys);

    num = numKeys();
    // delete any keys outside the range 
    for (i = 0; i < num; i++) {
	val = hashval(keys[i]);
	deleteKey = false;
	zscore = ((float) val - x1) / sigma;

	if (uflag) {
	    if (fflag && val > upper)
		deleteKey = true;
	    if (nflag && norm_cdf(zscore) > upper)
		deleteKey = true;
	    if (eflag && evd_cdf(alpha, beta, val) > upper)
		deleteKey = true;
	}
	if (lflag) {
	    if (fflag && val < lower)
		deleteKey = true;
	    if (nflag && norm_cdf(zscore) < lower)
		deleteKey = true;
	    if (eflag && evd_cdf(alpha, beta, val) < lower)
		deleteKey = true;
	}

	if (!deleteKey)
	    hashAssign(keys[i], 1);
	else
	    hashDel(keys[i]);

    }

    // grab updated set of keys
    hashKeys(&keys);
    rewind(fp);
    d = 0;
    while (d != EOF) {
	while ((d = fscanf(fp, "%s %u%[\r\n]", s, &val, c)) != EOF) {
	    s[Length] = '\0';
	    if (hashval(s)) {
		hashAssign(s, val);
	    }

	    if (d == 3)
		break;
	}
	if (d == 3) {
	    if (kflag) { 
	    	for (i = 0; i < numKeys()-1; i++) {
			printf("%s\t%u\t",keys[i], hashval(keys[i]));
			hashAssign(keys[i], 1);
	    	}
		printf("%s\t%u\n",keys[i], hashval(keys[i]));
		hashAssign(keys[i], 1);

	    } else {	    
	    	for (i = 0; i < numKeys()-1; i++) {
			printf("%u\t", hashval(keys[i]));
			hashAssign(keys[i], 1);
	    	}
	    printf("%u\n", hashval(keys[i]));
	    hashAssign(keys[i], 1);
	    }
	}
    }
}


// Add maximum word length given base function
